### PR TITLE
RFD 0179: Web API backward compatibility guidelines

### PR DESCRIPTION
**Pull request by avatus**
[rendered](https://github.com/gravitational/teleport/blob/07294183e9a197a952d79b41d729dd17a5c8fc3b/rfd/0179-api-versioning.md)

This RFD includes the API guidelines for keeping our web clients/api backward compatible

### CODE REVIEW

File: rfd/0179-api-versioning.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+## Solution
+
+**Summary**: Leverage ConnectRPC and create a web service that can use a proto schema to define our requests and response types. Also, optionally, use a modern fetching framework to simplify and unify the way we make api requests in the web client. This will remove boilerplate that we need to use our APIs in the web client, provide type safety out of the box for both server and client code, and provide backward compatibility guardrails when creating/updating our APIs. 
+
+
+### Creating a new rpc service with Connect
+(buf's Connectrpc, not Teleport Connect)
+
+Connect is a library used in making gRPC compatible HTTP apis. It supports [multiple protocols](https://connectrpc.com/docs/introduction#seamless-multi-protocol-support)  and can generate go service code and typescript client code from the proto schema. We do something similar in Teleport Connect where we generate a client for `tsh` to make requests to the server and use the generated typescript code. This would be doing the same for the web client now.
```
Comment (bl-nero) [inline rfd/0179-api-versioning.md:62]: Support for multiple protocols sounds great, but can we disable unused ones to narrow down attack vectors?

**Review by bl-nero — COMMENTED** [review]
My opinion is that incrementing a version number would be a perfectly good solution for some one-off breaking changes. There is definitely a big appeal in converting to Connect RPC, I can't deny, and my inner self striving for nice, complete solutions, is all in favor. But we need to do it for valid reasons, and openly acknowledge it.

As a solution for the problem of manual bookkeeping of the client-side and server-side data structure definitions, this is definitely something that I would support. Hell, this really triggered me sometimes. However, I don't think I've seen enough evidence that this actually helps to solve the backwards compatibility problem. In my eyes, we would still need to use similar techniques to tell old from new; only now, we would get some guardrails in form of both "new" and "old" being defined as a part of the same type-safe structure. Which we could achieve anyway with a bit of manual work, without involving Connect RPC.

Perhaps I don't see it? Maybe it would be easier to understand if you elaborated exactly how leveraging Connect RPC would help in the case that you mentioned?

**Comment by bl-nero** [comment]
One more thing: if we decide to use Connect RPC, can we make sure that we are prepared for debugging the network traffic? (I've seen solutions that make it really difficult to see what's happening between the browser and the server, so it's important for this to be either human-readable or have a good tooling support.)

File: rfd/0179-api-versioning.md (hunk 1 above)
Comment (avatus) [reply rfd/0179-api-versioning.md:62]: Yes you can select your transport.

**Comment by avatus** [comment]
> One more thing: if we decide to use Connect RPC, can we make sure that we are prepared for debugging the network traffic? (I've seen solutions that make it really difficult to see what's happening between the browser and the server, so it's important for this to be either human-readable or have a good tooling support.)

yeah, this is human readable still 
https://github.com/gravitational/teleport/pull/44668/files#diff-7852f06643f466fb42c702c6f8bb51f67339dbb53e40629bce1c2a69918d1d0dR107

https://connectrpc.com/docs/go/getting-started#make-requests

**Comment by avatus** [comment]
> Which we could achieve anyway with a bit of manual work, without involving Connect RPC.
Thanks, I need to update the language of this PR (working on it now). This shouldn't be "hey here is an idea!" type of RFD, but a prescription of rules/guidelines. I am using Connect as a way to facilitate these changes, but the guidelines I'm writing up (will be in tomorrow) will tell us how to keep our APIs backward compatible (with manual work) and here is a tool to help us do it. 

thanks for the feedback.

at the end of the day, anyone can break anything if they try hard enough :)

**Comment by avatus** [comment]
Ready for review

**Comment by avatus** [comment]
@ravicious @gzdunek i know this RFD is for the web UI, but I'd still appreciate any input. Especially on the optional use of tanstack query

File: rfd/0179-api-versioning.md [hunk 2]
```diff
@@ -51,15 +68,195 @@ This may work, but it highlights a couple things.
...
+fulfilled and NOT before. This means marking for deletion with a TODO and
+removing it in the relevant version. For example, if v17 no longer uses
+`GET /webapi/foos` , you can mark for deletion in v18, as v18 only needs to
+support n-1 versioned clients. It is the responsibility of the developer who is
+"removing the use" (i.e., making it obsolete with some other endpoint) to mark
+the TODO.
+
+
+### Adding/removing a field to a response of an existing endpoint
+Adding a new field to a response is OK as long as that field has to effect on
```
Comment (bl-nero) [inline rfd/0179-api-versioning.md:117]:
```suggestion
Adding a new field to a response is OK as long as that field has no effect on
```

File: rfd/0179-api-versioning.md [hunk 3]
```diff
@@ -51,15 +68,195 @@ This may work, but it highlights a couple things.
...
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we must keep ContainsDuplicates populated to support older clients!"
+	     ContainsDuplicates: true
+	     DuplicateCount: 1 
+	}, nil
+}
+``` 
+If using ConnectRPC (a proposed solution later in this RFD), then we must _not_ remove fields from requests/responses, even if unused. Mark as deprecated and move on.
```
Comment (bl-nero) [inline rfd/0179-api-versioning.md:166]: Do you mean that marking it `reserved` won't work?

Comment (avatus) [reply rfd/0179-api-versioning.md:166]: reserved will work. I should change the language to "follow the deprecation process of proto messages and move on". i will update and link. Thanks!

**Review by ravicious — COMMENTED** [review]
I'll have a couple of more thoughts on this that I'll post later today.

File: rfd/0179-api-versioning.md [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+}
+```
+
+This breaks backward compatibility because a web client of version n-1 would be
+expecting a list of requests only, and a web client of n hitting a proxy of n-1
+would be expecting to receive the paginated list.
+
+The solution _in this particular case_ was
+1. on the proxy side, check which version of the web client was making the request (by the existence of a specific query param, `limit`) and decide which shape to send
+2. on the web client side, if receiving a response from an old proxy, fallback to the client pagination.
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:62]: Yeah this one is a weird one, but definitely a valid concern. A few customers of ours specifically have run into issues running multiple versions of the proxy in a rolling upgrade, which causes major issues for them when the web client (served by proxy N) hits a proxy N-1 in their load balancer. 

It's one of the reasons for the JavaScript bundle condense project earlier this year because web clients couldn't even download the right bundle files properly when split. This happened with API calls as well

It's essentially an edge case, but large enough to consider.

File: rfd/0179-api-versioning.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+This may work, but it highlights a couple things.
+- we can send any response shape we want with no errors/warnings. 
+- if backward compatibility wasn't forefront of the mind, there was nothing stopping us from just pushing the new code with a break
+- the backend has to check the request shape and the frontend has to check the response shape just to know what type it is receiving.
+
+Using a standard guideline for an rpc service should help us evolve our API overtime in a backward compatible way.
+
+## Web Client/API Backward Compatibility Guidelines
+
+In general, if the Request or Response shape does not change, logic in an API handler can change without worry of backward compatibility.
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:73]: The shape of request or response doesn't change so the API contract is still fulfilled. This means it wouldn't lead to a "breaking" change

I guess a contrived example would be changing to return value in a response like {"hello": "world"} to {"hello": ", world!"} wouldn't be breaking. 

But after typing it out, perhaps some logic is done by that field in the web client so. I'm not sure now actually. I'll think about it

File: rfd/0179-api-versioning.md [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+would not cause a break (i.e., an old client sending the old request without the
+new field). If the handler _requires_ this new field in order to function, then
+a new endpoint must be created with a version increment.
+
+
+### Adding new features to the web client
+If a feature requires a new endpoint, it must properly handle the case of a 404.
+This means making the "expected" 404 error clear to the user that an endpoint is
+not found due to a version mismatch (and preferably which minimum version is
+needed) rather than an ambiguous "Not Found" error.
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:181]: Good point. I think in general the sentiment stands but perhaps I can word it differently. I would say in most cases a List returning a 404 would probably indicate a missing endpoint rather than a single resource (I don't think there is much we can do here besides still returning some shape?)

File: rfd/0179-api-versioning.md [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+
+An example of this in recent memory is Pinned Resources. The "pinned resources"
+tab was available in clients that _may_ have been communicating with proxies
+older than the required version. So a check was made to see if that proxy
+supported pinned resources, and if it wasn't, made the error known to the user
+that "Pinned Resources is supported by proxies version 14.1+" (or something like
+that).
+
+### Updating request fields in the web client
+Updating what fields are sent in a request body is ok as long as the API follows the updated request fields guidelines.
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:200]: Yeah basically just the direction of the communication. I have them separated since I went with "server changes" and "web client changes". I think that you were able to assume the correct answer that it makes _enough sense_ but I will reword it to make it more clear regardless (or note or something). Thank you!

File: rfd/0179-api-versioning.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:181]: Also I guess this would be up to the discretion of the dev at time of creation. Definitely one to just keep an eye out for (not a _strict_ requirement but a guideline)

A good example of this was Pinned Resources. We would check if the endpoint existed and if not, display a message informing users to upgrade to 14.1 or something like that. I could have swore I had that example in here but I will add that example for clarity

[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:181]: Leaf servers also play a big role in this one (and the previous comment about rolling upgrades too)

File: rfd/0179-api-versioning.md [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+automatically.
+
+### Add a new RPC service for the web API
+
+**Summary**: Leverage ConnectRPC and create a web service that can use a proto
+schema to define our requests and response types. Also, optionally, use a modern
+fetching framework to simplify and unify the way we make api requests in the web
+client. This will remove boilerplate that we need to use our APIs in the web
+client, provide type safety out of the box for both server and client code, and
+provide backward compatibility guardrails when creating/updating our APIs.
```
Comment (ravicious) [inline rfd/0179-api-versioning.md:248]: I totally agree with @bl-nero's comment above about bumping a version number being good enough for one-off breaking changes and the biggest advantage of Connect RPC being not forwards/backwards compatibility, but autogenerated type-checked structures in both JS and Go.

As far as compatibility is concerned, I feel like Connect RPC itself doesn't provide anything which we couldn't replicate with something like, idk, zod on the JS side. The difference is that with zod we would need to replicate the same structures in JS and then in Go.

I don't think we'll find anyone who, after reading this RFD, would be able to conclude with "This is a great that will absolutely work and be worth the time, let's do it". Perhaps it would be worthwhile to do a spike with an example implementation to see:

* How easy it'll be to integrate this along the current web API.
* How Connect RPC would address common pain points or past problems compared to the existing solution.

---

Another idea to toy with which could turn out to be an argument in favor of RPCs: how many web API endpoints do actual work and how many of them merely send off data to the auth service? Given the unusual compatibility guarantees put upon the Web UI (a client that needs to work with older servers), how feasible would it be to reuse protos from the `/api` directory to streamline endpoints which merely pass data to the auth service?

By the last question, I mean that when I want to add a new field to a usage event in Connect, which passes from the Electron app through tshd to prehog, it's a matter of updating a single proto file and the change trickles down to every place that uses it. Is this something achievable in the Web UI as well?

File: rfd/0179-api-versioning.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply rfd/0179-api-versioning.md:62]: It might be worth it to explicitly state it out in the RFD, since it's not a compatibility guarantee that's typically thought about in the context of any other client <-> server relationship within Teleport.

File: rfd/0179-api-versioning.md (hunk 8 above)
Comment (avatus) [reply rfd/0179-api-versioning.md:248]: > The difference is that with zod we would need to replicate the same structures in JS and then in Go.

This is the crux of the problem. At the end of the day, if we rely on creating code on the backend _and_ the frontend, we are bound to just end up in the same place we are. Removing the need to create code in two places and removing boilerplate seems worthwhile. A spike is probably a great idea (more than I spike i did for implementing a test feature, more like spike an actual feature instead). 

>  how feasible would it be to reuse protos from the /api directory to streamline endpoints which merely pass data to the auth service?

Probably pretty feasible. Although, from my experience, we send some fields that _arent_ passed to auth (and are used for creating the auth message generally) but its the return trip where things are quite different. Almost every resource-based web endpoint receives data from auth an has to convert it to a UI structure so reusing protos from proxy<->auth doesn't really seem to work.
I think the benefit that Connect gives us is the type generation is a bit more modern in that it reads closer to how types would be generated by a human, and also (if we went the tanstack query route), has built in generators for client queries as well. A lot it the type system _could_ be solved without it but from my initial testing, this was the best DX imo. 

I agree that solving the backward compatibility problem is as simple as appending the version prefix I talked about in the RFD. If we want to discuss the type system in a greater context, we can move it to its own RFD.

Comment (kimlisa) [reply rfd/0179-api-versioning.md:248]: > If we want to discuss the type system in a greater context, we can move it to its own RFD.

+1

File: rfd/0179-api-versioning.md [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+``` 
+If using ConnectRPC (a proposed solution later in this RFD), then we must _not_ remove fields from requests/responses, even if unused. Mark as deprecated and move on.
+
+If a response shape must change entirely, prefer creating a new endpoint/RPC.
+
+### Updating Request body fields
+A Request body can receive a new field as long as the omission of the new field
+would not cause a break (i.e., an old client sending the old request without the
+new field). If the handler _requires_ this new field in order to function, then
+a new endpoint must be created with a version increment.
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:174]: The request body could be something like this
```go
type MyRequest struct {
    Limit int
    Query string
}
```

And if we wanted to add a new field like
```go
type MyRequest struct {
    Limit int
    Query string
    MyNewThing string
}
```

if the handler has something along the lines of
```go
if req.MyNewThing == "" {
     return nil, err("BAD REQUEST")
}
```
that would be considered a breaking change because old clients do NOT send `MyNewThing` with their requests. therefore, a new endpoint is needed.

But, similar to the paginated access requests, we didn't need to change the endpoint because we added a new value `Limit` to the request, and if it wasn't present, we just considered it not paginated. This _wouldn't_ be a breaking change and should be allowed.

File: rfd/0179-api-versioning.md [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+### Updating request fields in the web client
+Updating what fields are sent in a request body is ok as long as the API follows the updated request fields guidelines. 
+
+---
+### A short checklist when updating the web client/api
+- [ ] If a new endpoint was added to accommodate a new feature, the new feature must fail gracefully with the required proxy version
+- [ ] When updating a feature, check if the web client works as expected with a proxy version N-1
+- [ ] When updating a feature, check that the proxy works receiving requests from web clients N-1
+- [ ] If deprecating an endpoint, a TODO is added to delete in the relevant version
+- [ ] When updating an API response, deprecated fields are still populated
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:208]: Thats a great idea!!

File: rfd/0179-api-versioning.md [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+
+---
+
+## Solution
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:212]: Hmmm, I suppose you're right. I'll switch it around. I'd still like the Connect RPC part to be at the bottom as it's an optional upgrade.

File: rfd/0179-api-versioning.md [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+#### Incrementing version prefix
+We should [stop stripping the api prefix](https://github.com/gravitational/teleport/blob/master/lib/web/apiserver.go#L527-L530), and add a new endpoint with a incremented prefix to accommodate the new response shape.
+```go
+
+// regex to match /v1, /v2, /v3, etc
+apiPrefixRegex := regexp.MustCompile(`^/v(\d+)/webapi`)
+// request is going to the API?
+if matches := apiPrefixRegex.FindStringSubmatch(r.URL.Path); matches != nil {
+    version := matches[1]
+    // Strip the prefix only if it's v1
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:223]: > what would be the benefit of stripping it?
just to keep it "as is" and not have to append `/v1` to all of our existing handlers in the backend.

File: rfd/0179-api-versioning.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:200]: Thank you for the feedback. I will update! 🙏

File: rfd/0179-api-versioning.md [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+	     Foos: foos,
+	     ContainsDuplicates: true
+	}, nil
+}
+```
+
+### Removing an endpoint
+An endpoint can be removed after our backward compatibility promise has been
+fulfilled and NOT before. This means marking for deletion with a TODO and
+removing it in the relevant version. For example, if v17 no longer uses
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:109]: I don't think minors count? As far as I can tell, we only worry about [major versions](https://goteleport.com/docs/upgrading/overview/#component-compatibility) when it comes to declaring incompatibility. Why does something in 17.1 have to stay until 19 compared to something in 17.0 being ok to be gone in 18?

File: rfd/0179-api-versioning.md [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+`GET /webapi/foos` , you can mark for deletion in v18, as v18 only needs to
+support n-1 versioned clients. It is the responsibility of the developer who is
+"removing the use" (i.e., making it obsolete with some other endpoint) to mark
+the TODO.
+
+
+### Adding/removing a field to a response of an existing endpoint
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:116]: Yes, great idea

File: rfd/0179-api-versioning.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:212]: In this context, "solution" meant "how can we implement these guidelines I'm bringing up because we currently don't have a way" but I'm going to reword it a bit

**Review by flyinghermit — APPROVED** [review]
Reviewed only the versioning part and that LGTM.

The api versioning guideline and the idea of backward compatibility itself is good, though it feels unfortunate that we have to deal with this situation of proxy running with different version in a same cluster in the first place. Alternatively, can we make changes in the proxy so it would always forward incoming request to the proxy which has the latest API version? That way we wouldn't need to worry about internal req/resp fields as long as the request hits correct proxy.

The idea of using Connect RPC sounds good to me and I see single type system as a win. Maybe worth it to move it to a different RFD as suggested in existing review comments so we have more discussion on that.

File: rfd/0179-api-versioning.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (kimlisa) [reply rfd/0179-api-versioning.md:109]: this is the scenario i was trying to describe: if something got introduced `v17.1`, then you have to mark for deletion in `v19` (not `v18`), b/c if we delete in v18, then we break compatibility with v17.0 -> v18.0

am i misunderstanding? i do get confused by this all the time so maybe i am 😅

[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:109]: Oh I see what you are saying. Then yes YOU are correct and I will update the example!

[the remark this answers is not in the document]
Comment (kimlisa) [reply rfd/0179-api-versioning.md:109]: oh btw, i think we need to include the `patch` too... probably unlikely to happen but it could happen, so:

- new feature introduce on major version, mark deletion for one major version ahead
- new feature introduce on minor/patch version, mark deletion for two major version ahead

File: rfd/0179-api-versioning.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (gzdunek) [reply rfd/0179-api-versioning.md:62]: >A few customers of ours specifically have run into issues running multiple versions of the proxy in a rolling upgrade, which causes major issues for them when the web client (served by proxy N) hits a proxy N-1 in their load balancer.

I'm thinking how versioning will help with this. Let's assume the example from above. The developer adds a new `v2/webapi/accessrequests` endpoint that returns a paginated list of resources. This endpoint is called by the new version of the frontend. Now, when a rolling update is in progress, we may have two versions of the proxy. What should happen if we hit the older proxy with the new frontend and get 404? 
- Show the error to the user? But this means that we will end up with a broken service (I know that you mentioned handling 404 in the other paragraph, but that one was about a new feature, no?)
- Retry the call but to `v1/*` and then do that manual conversion to the paginated list on the JS side? 

Probably the ideal solution for such cases would be to have a load balancer that routes the requests based on the required version. Unfortunately, I don't think it's something we can require from customers (+ configuring this would be non-trivial) :(

File: rfd/0179-api-versioning.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (gzdunek) [reply rfd/0179-api-versioning.md:73]: >The shape of request or response doesn't change so the API contract is still fulfilled. This means it wouldn't lead to a "breaking" change

I agree that it's not true. 
For example, if an API returns `status: "SUCCES"`, and then we change it to `status: "OK"` it is a breaking change.  

I think we should say instead that meaning of the fields should not change? Like in https://www.mnot.net/blog/2012/12/04/api-evolution.html (paragraph Make Changes Backwards-Compatible).

**Review by gzdunek — COMMENTED** [review]
Left some comments.
In general, I agree with others that ConnectRPC is a really good improvement but doesn't change much when it comes to API compatibility (so we can discuss it separately).

File: rfd/0179-api-versioning.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply rfd/0179-api-versioning.md:109]: Wow, yeah, it didn't occur to me, but yeah, it's something we do have to account for.

Perhaps we could explain the general rule and then show examples, rather than adding one rule and then being like "oh in case a feature is added in a minor or patch version, there's a slightly different rule". I was thinking of something like this:

> An endpoint can be removed in a major version n+2, where n is the last major version where the endpoint was used.
>
> Example 1: v17 no longer uses `GET /webapi/foos` which was last used in v16. The endpoint can be removed in v18.
> Example 2: v4.2.0 no longer uses `GET /webapi/bars` which was last used in v4.1.3. The endpoint can be removed in v6, so that v5 still supports clients using v4.1.3 and v4.2.0.

File: rfd/0179-api-versioning.md [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+Adding a new field to a response is OK as long as that field has to effect on
+the previously existing fields. Any field that was previously in a response
+should _stay populated_ in the new response, even if that creates duplicate
+data. An existing field _cannot_ have its type changed. A field should not be
+removed until the backward compatibility promise has been 'fulfilled'. If v17
+clients are only using a new field, you can mark the old field as "TODO DELETE"
+in v18. This means that v17 servers can support v16 clients (still using the old
```
[the remark this answers is not in the document]
Comment (ravicious) [reply rfd/0179-api-versioning.md:123]: The more popular format is `TODO(<github-handle>): DELETE IN 18.0`, without the `v`. 🤓

```
$ g show-ref HEAD
30f3deb731d511b355592ce23b22a65dd294f9c7 refs/remotes/origin/HEAD
$ rg "DELETE IN v" | wc -l
      25
$ rg "DELETE IN \d" | wc -l
     155
```

File: rfd/0179-api-versioning.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:62]: > Show the error to the user? But this means that we will end up with a broken service (I know that you mentioned handling 404 in the other paragraph, but that one was about a new feature, no?)

New feature yes but in this scenario it also works. Think of leaf clusters as well, they may be behind a version or two and could have the same issue (such as pinned resources). Most users don't do this rolling upgrade, so they won't see it. and I think a feature (either updated or new) showing a message saying "hey this doesn't work correctly until its all upgraded" isn't a bad thing. It lets the users know "teleport isn't randomly broken, its broken because of _this_" is better than random 404 and also better than trying to call an older version of a web api which, after these guidelines, is only there because it could _not_ support this new update.

[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:62]: > Unfortunately, I don't think it's something we can require from customers

yeah agreed. it was an idea we had when we were trying to solve the app.js bundle issue and realized it was too out of our hands

File: rfd/0179-api-versioning.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:73]: Ok I'm convinced. I will update this!

File: rfd/0179-api-versioning.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:181]: I've added the distinction between "list" endpoints and single resource endpoints returning 404. I believe its going to be "Case-by-case" but the guidelines should be "best effort to handle gracefully"

File: rfd/0179-api-versioning.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:200]: I've removed this second paragraph and slightly changed the wording on the other to include both sides of the change.

File: rfd/0179-api-versioning.md [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+"removing the use" (i.e., making it obsolete with some other endpoint) to mark
+the TODO.
+
+
+### Adding/removing a field to a response of an existing endpoint
+Adding a new field to a response is OK as long as that field has to effect on
+the previously existing fields. Any field that was previously in a response
+should _stay populated_ in the new response, even if that creates duplicate
+data. An existing field _cannot_ have its type changed. A field should not be
+removed until the backward compatibility promise has been 'fulfilled'. If v17
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:121]: I've removed that line and linked to the guidelines for removing an endpoint (since they should be the same guidelines for fields as well)

File: rfd/0179-api-versioning.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:174]: Added this explaination snippet to the RFD

File: rfd/0179-api-versioning.md [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+a new endpoint must be created with a version increment.
+
+
+### Adding new features to the web client
+If a feature requires a new endpoint, it must properly handle the case of a 404.
+This means making the "expected" 404 error clear to the user that an endpoint is
+not found due to a version mismatch (and preferably which minimum version is
+needed) rather than an ambiguous "Not Found" error.
+
+### New required fields from an API response in the web client
```
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:183]: This is about the response shape. If the web client expects some _new_ field in the response, it needs to be calling a new version.

**Comment by avatus** [comment]
Thanks for all the feedback team. I've removed the talk about adding Connect (im still excited on this so I will open another RFD soon). Also, I believe I've handled all the feedback. Please take another look 🙏 🙌

**Comment by avatus** [comment]
Any lingering feedback on this? Friendly ping to reviewers

File: rfd/0179-api-versioning.md [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+"resources" endpoint where it is expected to return an empty list if something
+isn't found (i.e., "no access requests"), but a new endpoint that returns a
+single resource _should not_ assume that a 404 means anything other than a
+resource is not found. In this particular case, there isn't much graceful work
+we can do here. Graceful handling of this error state should be best effort.
+
+### New required fields from an API response in the web client
+If the updated feature _cannot_ function without receiving new fields from the
+API (for example, receiving a response from a proxy version N-1), refer to the
+API guidelines about creating a new endpoint. If the feature itself is degraded
```
Comment (gzdunek) [inline rfd/0179-api-versioning.md:247]: Nit: the guideline you are referring to is about creating an endpoint that didn't exist before, not about the version increase.

**Review by gzdunek — COMMENTED** [review]
Sorry for the delay, I completely missed that you updated the RFD last week :(

File: rfd/0179-api-versioning.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (avatus) [reply rfd/0179-api-versioning.md:244]: I believe this would be handled with the `Updating request body fields` section. I will add your example and a disclaimer about new version should be a default!

File: rfd/0179-api-versioning.md [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+```
+
+### Removing an endpoint
+An endpoint can be removed in a major version n+2, where n is the last major
+version where the endpoint was used.
+
+Example 1: v17 no longer uses GET /webapi/foos which was last used in v16. The
+endpoint can be removed in v18. 
+Example 2: v4.2.0 no longer uses GET /webapi/bars which was last used in v4.1.3. The endpoint can be removed in v6,
+so that v5 still supports clients using v4.1.3 and v4.2.0.
```
Comment (kimlisa) [inline rfd/0179-api-versioning.md:156]: it wasn't separated (run on)
```suggestion
Example 1: v17 no longer uses GET /webapi/foos which was last used in v16. The
endpoint can be removed in v18. 

Example 2: v4.2.0 no longer uses GET /webapi/bars which was last used in v4.1.3. The endpoint can be removed in v6, so that v5 still supports clients using v4.1.3 and v4.2.0.
```

File: rfd/0179-api-versioning.md [hunk 20]
```diff
@@ -0,0 +1,392 @@
...
+    // TODO (avatus) DELETE IN 18
+    ContainsDuplicates bool
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we must keep ContainsDuplicates populated to support older clients!"
```
[the remark this answers is not in the document]
Comment (kimlisa) [reply rfd/0179-api-versioning.md:200]: i didn't see this change 😅

File: rfd/0179-api-versioning.md [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+"resources" endpoint where it is expected to return an empty list if something
+isn't found (i.e., "no access requests"), but a new endpoint that returns a
+single resource _should not_ assume that a 404 means anything other than a
+resource is not found. In this particular case, there isn't much graceful work
+we can do here. Graceful handling of this error state should be best effort.
+
+### New required fields from an API response in the web client
+If the updated feature _cannot_ function without receiving new fields from the
+API (for example, receiving a response from a proxy version N-1), refer to the
+API guidelines about creating a new versioned endpoint. If the feature itself is degraded
```
Comment (kimlisa) [inline rfd/0179-api-versioning.md:254]:
```suggestion
API (for example, receiving a response from a proxy version N-1), refer to the
API guidelines about [creating a new versioned endpoint](#creating-a-new-endpoint). If the feature itself is degraded
```

File: rfd/0179-api-versioning.md [hunk 22]
```diff
@@ -0,0 +1,265 @@
+authors: Michael Myers (michael.myers@goteleport.com)
+state: draft
+---
```
Comment (kimlisa) [inline rfd/0179-api-versioning.md:3]: think you need the extra hyphens on top to be formatted correctly

```suggestion
---
authors: Michael Myers (michael.myers@goteleport.com)
state: draft
---
```

File: rfd/0179-api-versioning.md [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+	     Foos: foos,
+	     ContainsDuplicates: true
+	}, nil
+}
+```
+
+### Removing an endpoint
+An endpoint can be removed in a major version n+2, where n is the last major
+version where the endpoint was used.
+
```
Comment (kimlisa) [inline rfd/0179-api-versioning.md:152]: think its good add a delete example:

``````suggestion

Mark endpoints that needs to be removed with:
```go
// TODO(<your-github-handle>): DELETE IN 18.0
h.GET("/webapi/tokens", h.WithAuth(h.getTokens))
```

``````

File: rfd/0179-api-versioning.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (kimlisa) [reply rfd/0179-api-versioning.md:251]: omg, i just now saw/understood the difference. `response` vs `request` 🤦‍♀️

**Review by kimlisa — COMMENTED** [review]
LGTM, i'll leave approving to @kiosion since things are new for him and can request any clarification

File: rfd/0179-api-versioning.md [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+// keep this
+h.GET("/webapi/tokens", h.WithAuth(h.getTokens))
+
+// and add this when a new version
+h.GET("/v2/webapi/tokens", h.WithAuth(h.getTokens))
+```
+
+## Web Client/API Backward Compatibility Guidelines
+
+A proxy server verison N must support a web client of version N and N-1. A web client of version N _might_ hit a proxy of version
```
Comment (rudream) [inline rfd/0179-api-versioning.md:98]:
```suggestion
A proxy server version N must support a web client of version N and N-1. A web client of version N _might_ hit a proxy of version 
```

**Review by kimlisa — APPROVED** [review]
rest looks like minor comments, approving

**Review by rudream — APPROVED** [review]
